### ISSUE TITLE
import.meta.dirname and import.meta.filename from Node.js 20.11.0 not present in webpack bundled page

### ISSUE THREAD

**Issue by karlhorky** [vercel/next.js#60879]
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/import-meta-dirname-reproduction-vj3gfq?file=%2Fapp%2Fpage.tsx%3A1%2C1

### To Reproduce

1. Open the reproduction
2. Observe `import.meta.dirname` is `undefined` in the webpack-compiled `page.tsx` (but is a dir path in Node.js)

![Screenshot 2024-01-19 at 11 45 54](https://github.com/vercel/next.js/assets/1935696/575bea38-1960-49fa-bf4a-b07ce45a78e4)

Changes to repro template:

- switch to `"type": "module"` to enable running ESM with Node.js
- add file `import-meta.dirname.js` with `console.log()` of `import.meta.dirname`
- add `console.log()` of `import.meta.dirname` to `app/page.tsx`

### Current vs. Expected behavior

I expected [`import.meta.dirname` and `import.meta.filename`](https://github.com/nodejs/node/pull/48740), introduced in [Node.js 20.11.0](https://nodejs.org/en/blog/release/v20.11.0) to not be `undefined` in Next.js pages in the App Router, but it is `undefined`

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP PREEMPT_DYNAMIC Sun Aug  6 20:05:33 UTC 2023
Binaries:
  Node: 20.11.0
  npm: 10.2.4
  Yarn: 1.22.19
  pnpm: 8.14.1
Relevant Packages:
  next: 14.1.1-canary.1
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.1.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Module resolution (CJS / ESM, module resolving)

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local)

**Comment by karlhorky** [comment]
cc @alexander-akait if I should open an issue in https://github.com/webpack/webpack, let me know and I'll do that.

**Comment by alexander-akait** [comment]
@karlhorky Yeah, to be honestly it is already in my internal roadmap :smile:

**Comment by karlhorky** [comment]
## Workaround

Configure the `DefinePlugin` to alias `import.meta.dirname` to `__dirname`, similar to https://github.com/webpack/webpack/issues/1599#issuecomment-186620333:

```js
/** @type {import("next").NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  webpack: (
    config,
    { buildId, dev, isServer, defaultLoaders, nextRuntime, webpack },
  ) => {
    // Workaround for webpack not supporting import.meta.dirname
    // - https://github.com/vercel/next.js/issues/60879#issuecomment-1900215814
    config.plugins.push(
      new webpack.DefinePlugin({
        "import.meta.dirname": "__dirname",
      }),
    );
    return config;
  },
};

export default nextConfig;
```

CodeSandbox demo: https://codesandbox.io/p/devbox/import-meta-dirname-reproduction-forked-4ldmgw?file=%2Fnext.config.js%3A3%2C25

![Screenshot 2024-01-19 at 12 14 13](https://github.com/vercel/next.js/assets/1935696/ff33c8c6-d3df-48ef-a9a8-b5622975398d)

**Comment by huozhi** [comment]
@alexander-akait let us know when it's landed in webpack, would love to help bump webpack in next.js for this later 🙏

**Comment by kachkaev** [comment]
How hard would it be to add @karlhorky’s [workaround](https://github.com/vercel/next.js/issues/60879#issuecomment-1900215814) in the meantime? I just upgraded to Node 20, used `import.meta.dirname` and it worked in unit tests. TypeScript linter did not complain as well, so I committed and pushed. CI failed because of this issue and it took me a few minutes to deal with it. Next.js is about DX, so it’d be great to see `import.meta.dirname` / `import.meta.filename` to Just Work™

**Comment by kachkaev** [comment]
Looks like turbopack does not support `import.meta.dirname` and `import.meta.filename` too. It returns undefined. And because webpack config is ignored, we can no longer leverage `webpack.DefinePlugin({...})` as a workaround. It’d be cool to use these two little shorthands because they are natively supported by Node 20.

**Comment by karlhorky** [comment]
Ah wait, looks like there maybe is one already:

- https://github.com/webpack/webpack/issues/18320

**Comment by alexander-akait** [comment]
@karlhorky Yeah, it is an issue about it

**Comment by karlhorky** [comment]
Looks like this PR closes it:

- https://github.com/webpack/webpack/pull/20050

@alexander-akait will this be released in the next minor?

- next minor: [`webpack@5.103.0`](https://github.com/webpack/webpack/releases/tag/v5.103.0)

**Comment by thernstig** [comment]
Looks like turobpack still does not support them?
